### PR TITLE
Update to Vs2019

### LINKED ISSUE #261: [REQUEST] Upgrade to VS2019
**Is your feature request related to a problem? Please describe.**
I wish to develop under VS2019 (or newer). Current set up process does not allow this.

**Describe the solution you'd like**
Amend CMakeLists as necessary to support VS2019. Do not forget to adapt BlueFramework as well!

### LINKED ISSUE #308: [REQUEST] Replace throw with noexcept
**Is your feature request related to a problem? Please describe.**
New coding practice states:
- use `noexcept(false)` instead of `throw(...)` and
- use `noexcept(true)` instead of `throw()`

with function signatures.

**Describe the solution you'd like**
Replace everything in our code.

See also https://en.cppreference.com/w/cpp/language/noexcept_spec .

### LINKED ISSUE #413: [REQUEST] Program architecture graphic
**Is your feature request related to a problem? Please describe.**
A clear overview of the general architecture of the solution is missing.

**Describe the solution you'd like**
A graphic that entails the base blocks of the program, what functionality it provides, to help newcomers get an overview.

**Additional context**
See also: https://github.com/openjournals/joss-reviews/issues/3061#issuecomment-803536122

### PR DESCRIPTION
**Pull request by hechth**
Updated code to compile with VS2019. Resolves #261 .

Fixed bugs in templates and fixed some include paths. Added `UNICODE` compile definition alongside `_UNICODE` as carve needs the first one and MSVC the second.

Fixes #308 for geometry converter classes.

Fixes #413

### CODE REVIEW

File: CurrentlyExcluded/Infrastructure/Infrastructure.Net/Infrastructure.Net/obj/Debug/.NETFramework,Version=v4.0.AssemblyAttributes.cs [hunk 1]
```diff
@@ -0,0 +1,4 @@
+// <autogenerated />
+using System;
+using System.Reflection;
+[assembly: global::System.Runtime.Versioning.TargetFrameworkAttribute(".NETFramework,Version=v4.0", FrameworkDisplayName = ".NET Framework 4")]
```
Comment (pjanck) [inline CurrentlyExcluded/Infrastructure/Infrastructure.Net/Infrastructure.Net/obj/Debug/.NETFramework,Version=v4.0.AssemblyAttributes.cs]: What are autogenerated files doing here?

File: CurrentlyExcluded/Infrastructure/Infrastructure.Net/Infrastructure.Net/obj/Release/.NETFramework,Version=v4.0.AssemblyAttributes.cs [hunk 2]
```diff
@@ -0,0 +1,4 @@
+// <autogenerated />
+using System;
+using System.Reflection;
+[assembly: global::System.Runtime.Versioning.TargetFrameworkAttribute(".NETFramework,Version=v4.0", FrameworkDisplayName = ".NET Framework 4")]
```
Comment (pjanck) [inline CurrentlyExcluded/Infrastructure/Infrastructure.Net/Infrastructure.Net/obj/Release/.NETFramework,Version=v4.0.AssemblyAttributes.cs]: Same: What are autogenerated files doing here?

**Review by pjanck — CHANGES_REQUESTED** [review]
Two files are questionable, otherwise LFTM.

I hope that merging this PR with others will not cause any issues.

**Comment by pjanck** [comment]
Also, this branch has conflicts with the `development` branch, these need to be resolved by merging `development` into it first.

File: CurrentlyExcluded/Infrastructure/Infrastructure.Net/Infrastructure.Net/obj/Debug/.NETFramework,Version=v4.0.AssemblyAttributes.cs (hunk 1 above)
Comment (pjanck) [reply CurrentlyExcluded/Infrastructure/Infrastructure.Net/Infrastructure.Net/obj/Debug/.NETFramework,Version=v4.0.AssemblyAttributes.cs]: Resolved fac87985cec1aba915e4bfb666d1ec2bdc0f1f34

File: CurrentlyExcluded/Infrastructure/Infrastructure.Net/Infrastructure.Net/obj/Release/.NETFramework,Version=v4.0.AssemblyAttributes.cs (hunk 2 above)
Comment (pjanck) [reply CurrentlyExcluded/Infrastructure/Infrastructure.Net/Infrastructure.Net/obj/Release/.NETFramework,Version=v4.0.AssemblyAttributes.cs]: Resolved fac87985cec1aba915e4bfb666d1ec2bdc0f1f34

File: CMakeLists.txt [hunk 3]
```diff
@@ -111,8 +111,11 @@ if(OIP_USE_PREDEFINED_THIRDPARTY_PATH)
 	find_path(OIP_Boost_DIR NAMES boost/version.hpp HINTS
 		"C:\\thirdparty\\${MSVC_VERSION_STRING}\\x64\\boost_1_75_0"
 		"C:\\thirdparty\\boost_1_75_0")
-	find_path(OIP_QT_DIR NAMES Qt5 HINTS "C:\\Qt\\5.14.2\\msvc${MSVC_VERSION_YEAR}_64\\lib\\cmake")
-	find_path(OIP_PROJ_DIR NAMES proj-config.cmake HINTS "C:\\OSGeo4W64\\apps\\proj-dev\\share\\cmake\\proj4\\proj")
+	find_path(OIP_QT_DIR NAMES Qt5 HINTS "C:\\Qt\\5.15.1\\msvc${MSVC_VERSION_YEAR}_64\\lib\\cmake")
+	find_path(OIP_PROJ_DIR NAMES proj-config.cmake HINTS "C:\\OSGeo4W\\share\\cmake\\proj4\\proj")
+	SET(CMAKE_INCLUDE_PATH ${CMAKE_INCLUDE_PATH} ${OIP_Boost_DIR})
+    SET(CMAKE_LIBRARY_PATH ${CMAKE_LIBRARY_PATH} "${OIP_Boost_DIR}\\lib")
```
Comment (pjanck) [inline CMakeLists.txt:109]: We need to make sure that these directories were found, before we add them to the `..._PATH`. Noting so we don't forget.

**Review by pjanck — CHANGES_REQUESTED** [review]
A note before we finish this.

**Review by pjanck — COMMENTED** [review]
Seems optimistic :)

**Comment by jschlenger** [comment]
Cmake creates "Automoc" folder in the solution.
This should be disabled.

**Review by jschlenger — COMMENTED** [review]
Review to better keep track of changes

File: CMakeLists.txt [hunk 4]
```diff
@@ -72,11 +71,17 @@ endif()
...
+	set(CMAKE_CXX_STANDARD 14)
+elseif(MSVC_VERSION GREATER_EQUAL 1920 AND MSVC_VERSION LESS_EQUAL 1929)
+	set(MSVC_VERSION_STRING "vs2019")
+	set(MSVC_VERSION_YEAR "2019")
+	set(CMAKE_CXX_STANDARD 20)
 else()
-	message("Please switch to Visual Studio 14 2015 or Visual Studio 15 2017.")
+	message("Please switch to Visual Studio 14 2015, Visual Studio 15 2017 or Visual Studio 16 2019")
 	return()
 endif()
```
Comment (pjanck) [inline CMakeLists.txt]: I believe that we'll be supporting VS2019 only.

File: Core/CMakeLists.txt [hunk 5]
```diff
@@ -71,7 +71,7 @@ target_link_libraries(OpenInfraPlatform.Core PRIVATE
 
 target_link_libraries(OpenInfraPlatform.Core PUBLIC OpenInfraPlatform.Base)
 
-if(MODULES_POINT_CLOUD_PROCESSING)
+if(OIP_WITH_POINT_CLOUD_PROCESSING)
 	add_definitions(-DOIP_WITH_POINT_CLOUD_PROCESSING)
```
Comment (pjanck) [inline Core/CMakeLists.txt:75]: perhaps use `add_compile_definitions` here as well?

File: Core/src/IfcGeometryConverter/SolidModelConverter.h [hunk 6]
```diff
@@ -1647,22 +1648,15 @@ namespace OpenInfraPlatform
...
-				else if (ifcBooleanOperator == typename IfcEntityTypesT::IfcBooleanOperator::ENUM::ENUM_INTERSECTION)
-				{
+				else if (typeid(IfcEntityTypesT::IfcBooleanOperator::ENUM::ENUM_INTERSECTION) == typeid(ifcBooleanOperator))
 					csgOperation = carve::csg::CSG::INTERSECTION;
-				}
-				else if (ifcBooleanOperator == typename IfcEntityTypesT::IfcBooleanOperator::ENUM::ENUM_DIFFERENCE)
-				{
+				else if (typeid(IfcEntityTypesT::IfcBooleanOperator::ENUM::ENUM_DIFFERENCE) == typeid(ifcBooleanOperator))
 					csgOperation = carve::csg::CSG::A_MINUS_B;
-				}
```
Comment (pjanck) [inline Core/src/IfcGeometryConverter/SolidModelConverter.h:1660]: These are really weird; but as long as it works ... :S

File: Documentation/markdown/ReleaseNotes.md [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+  * List with all supported Ifc entities (`#397 <https://github.com/tumcms/Open-Infra-Platform/pull/397>`_)
+
+  * Added detailed description of the software architecture (`#466 <https://github.com/tumcms/Open-Infra-Platform/pull/466>`_)
+
+
+Bug fixes
+---------
+
+* Fixing bug in alignment placement (`#443 <https://github.com/OSGeo/PROJ/issues/2692>`_)
```
Comment (pjanck) [inline Documentation/markdown/ReleaseNotes.md]: What's this link all about?

File: Documentation/markdown/SetupHelp.md [hunk 8]
```diff
@@ -21,71 +21,77 @@ Guide on how to compile the project TUM OpenInfraPlatform.
...
 		
-    * Open *Commandline_Utilities*
-    * Find *proj-dev* package 
-    * Click on *skip* to refresh the line.
+    * Open *Libs*
+    * Find *proj-devel* package 
+    * Click on *skip* to refresh the line and select version 8.1.1-1.
     * Select binary (*Bin*) package (the left column).
 		
 ![](../images/PROJ_Installation_settings.png)
```
Comment (pjanck) [inline Documentation/markdown/SetupHelp.md:92]: We should include some help for the point cloud support as well.

File: Documentation/markdown/SetupHelp.md [hunk 9]
```diff
@@ -21,71 +21,77 @@ Guide on how to compile the project TUM OpenInfraPlatform.
...
 
-5. **Boost 1_65_1** - find [here](https://sourceforge.net/projects/boost/files/boost-binaries/1.65.1/boost_1_65_1-msvc-14.1-64.exe/download).
+	* Create a folder `C:\thirdparty` and install **Boost 1_75_0** to `C:\thirdparty\vs2017\x64\boost_1_75_0` or `C:\thirdparty\vs2019\x64\boost_1_75_0`.
 
-	* Create a folder `C:\thirdparty` and install **Boost 1_65_1** to `C:\thirdparty\vs2017\x64\boost_1_65_1`.
-	* Add this path to the environment variables. (Create new environment variable called `Boost_INCLUDE_DIR`. This variable should point to the binary folder, where **Boost 1_65_1** is staged (e.g. `C:\thirdparty\vs2017\x64\boost_1_65_1`).
+*NOTE:* OIP expects *boost* at the path specified above. If you deviate from this, please set the `OIP_Boost_DIR` variable [correspondingly](./FAQ.md#libraries).
+
+	* Add this path to the environment variables. (Create new environment variable called `Boost_INCLUDE_DIR`. This variable should point to the binary folder, where **Boost 1_75_0** is staged (e.g. `C:\thirdparty\vs2017\x64\boost_1_75_0` or `C:\thirdparty\vs2019\x64\boost_1_75_0`).
```
Comment (pjanck) [inline Documentation/markdown/SetupHelp.md]:
```suggestion
	* Add this path to the environment variables. (Create new environment variable called `Boost_INCLUDE_DIR`. This variable should point to the binary folder, where **Boost 1_75_0** is staged (e.g. `C:\thirdparty\vs2019\x64\boost_1_75_0`).
```

File: Documentation/markdown/SetupHelp.md [hunk 10]
```diff
@@ -21,71 +21,77 @@ Guide on how to compile the project TUM OpenInfraPlatform.
...
 		
 ![](../images/Qt_Installation_settings.png)
 
-*NOTE:* There are components, which Qt Online Installer selects as default options. You can uncheck these components for saving computer memory.
+*NOTE:* There are components, which *Qt Online Installer* selects as default options. You can uncheck these components for saving computer memory.
+
+5. **Boost 1_75_0** - [vs2019](https://sourceforge.net/projects/boost/files/boost-binaries/1.75.0/boost_1_75_0-msvc-14.2-64.exe/download)
 
-5. **Boost 1_65_1** - find [here](https://sourceforge.net/projects/boost/files/boost-binaries/1.65.1/boost_1_65_1-msvc-14.1-64.exe/download).
+	* Create a folder `C:\thirdparty` and install **Boost 1_75_0** to `C:\thirdparty\vs2017\x64\boost_1_75_0` or `C:\thirdparty\vs2019\x64\boost_1_75_0`.
```
Comment (pjanck) [inline Documentation/markdown/SetupHelp.md]:
```suggestion
	* Create a folder `C:\thirdparty` and install **Boost 1_75_0** to `C:\thirdparty\vs2019\x64\boost_1_75_0`.
```

File: Documentation/markdown/SoftwareArchitecture.md [hunk 11]
```diff
@@ -141,6 +141,8 @@ The supported IFC versions are:
 * IFC4x2 (BIMROAD and DRAFT_1)
 * IFC4x3 (RC1 and RC4)
 
+It is relatively easy to introduce more recent IFC versions to OIP. All that is needed is the new schema definition in the form of the EXPRESS file. With this, the early binding generator creates C++ classes for all entities accordingly. Similar to the already implemented IFC versions, the OIP functionalities can then be modified based on the changes introduced through the new IFC version. With this, it is also possible to develop future IFC extensions and test their behaviour with low implementation effort.
```
Comment (pjanck) [inline Documentation/markdown/SoftwareArchitecture.md]: Please make these sentences reside in individual lines.

File: UserInterface/MainWindow.cpp [hunk 12]
```diff
@@ -1765,24 +1767,20 @@ void OpenInfraPlatform::UserInterface::MainWindow::on_actionShow_Help_triggered(
...
 
-#ifndef _DEBUG
-	filename = ss.str().append(L"log.txt");
+	ShellExecute(0, 0, ss.str().c_str(), 0, 0, SW_SHOW);
+#else
+	ShellExecute(0, 0, _T("log.txt"), 0, 0, SW_SHOW);
 #endif
 
-	ShellExecute(0, 0, filename.c_str(), 0, 0, SW_SHOW);
 }
```
Comment (pjanck) [inline UserInterface/MainWindow.cpp]: We aren't writing to log file anymore; we can remove the function alltogether.

**Review by pjanck — CHANGES_REQUESTED** [review]
This PR is getting quite huge ...

File: Core/src/IfcGeometryConverter/SolidModelConverter.h (hunk 6 above)
Comment (jschlenger) [reply Core/src/IfcGeometryConverter/SolidModelConverter.h:1660]: I wasn't able to find an example file that makes use of  Intersection, Difference or Union, but I can create an issue to check this part with an example file to see if it works properly.

File: Documentation/markdown/ReleaseNotes.md (hunk 7 above)
Comment (jschlenger) [reply Documentation/markdown/ReleaseNotes.md]: Done

File: CMakeLists.txt (hunk 4 above)
Comment (jschlenger) [reply CMakeLists.txt]: Done

File: Documentation/markdown/SetupHelp.md (hunk 9 above)
Comment (jschlenger) [reply Documentation/markdown/SetupHelp.md]: Done

File: Documentation/markdown/SoftwareArchitecture.md (hunk 11 above)
Comment (jschlenger) [reply Documentation/markdown/SoftwareArchitecture.md]: Done

File: Documentation/markdown/SetupHelp.md (hunk 10 above)
Comment (jschlenger) [reply Documentation/markdown/SetupHelp.md]: Done

File: UserInterface/MainWindow.cpp (hunk 12 above)
Comment (jschlenger) [reply UserInterface/MainWindow.cpp]: Done

**Pull request merged**